src/features/pn/services/cie/nisService.ts

```diff
@@ -0,0 +1,736 @@
+      );
+    }
+    consoleLogIfNeeded(
+      "Content of signed_attrs_set matches the expected value",
+      logDebugMessages
+    );
+
+    const signatureBytes = signatureFromASN1DER(asn1Der);
+    const signatureHex = bytesToUppercaseHex(signatureBytes);
+    if (signatureHex !== expectedSignatureHex) {
+      return commonFailureHandling(
+        `Content of signature DIFFERENT from expected\nExpected: ${expectedSignatureHex}\nComputed: ${signatureHex}\n`,
+        logDebugMessages
+      );
+    }
+    consoleLogIfNeeded(
+      "Content of signature matches the expected value",
+      logDebugMessages
+    );
+
+    const replacedSignedAttributesWrapperBytes = forge.util.hexToBytes(
+      canonizedSignedAttributesHex
+    );
+    const sha1ReplaceSignedAttributesMD = forge.md.sha1.create();
+    sha1ReplaceSignedAttributesMD.update(replacedSignedAttributesWrapperBytes);
+    const replacedSignedAttributesSHA1Bytes = sha1ReplaceSignedAttributesMD
+      .digest()
+      .getBytes();
+
+    const certificate = eIdentityCardSignerCertificateFromASN1DER(asn1Der);
+    const publicKey = certificate.publicKey as forge.pki.rsa.PublicKey;
+    if (!publicKey.verify(replacedSignedAttributesSHA1Bytes, signatureBytes)) {
+      return commonFailureHandling(
+        "Signature verification failed! The digital signature of the SOD is not valid",
+        logDebugMessages
+      );
+    }
+    consoleLogIfNeeded(
+      "Signature verification passed! The digital signature of the Document Security Object is valid",
+      logDebugMessages
+    );
+
+    return right(undefined);
+  } catch (error) {
+    const reason = unknownToString(error);
+    return commonFailureHandling(reason, logDebugMessages);
+  }
+};
+
+/**
+ * Verifies a challenge by decrypting a signature with an RSA public key and comparing
+ * the result to the original challenge. This function replicates the low-level
+ * behavior of `openssl rsautl -verify`.
+ *
+ * @param signedChallengeHex The RSA signature, as a hexadecimal string. This is the
+ * original challenge encrypted with the corresponding private key.
+ * @param publicKeyHex The RSA public key in DER format, encoded as a hexadecimal string.
+ * @param originalChallenge The original, unmodified challenge data, as a hexadecimal string,
+ * used for final comparison.
+ */
+export const verifyChallenge = (
+  signedChallengeHex: string,
+  publicKeyHex: string,
+  originalChallenge: string,
+  logDebugMessages: boolean = false
+): Either<string, void> => {
+  try {
+    // --- 1. Prepare the Public Key ---
+    // The public key is provided as a hex string representing its DER format.
+    // It must be converted into a node-forge public key object to be used.
+
+    // Convert the hex-encoded public key into a binary string.
+    const publicKeyDer = forge.util.hexToBytes(publicKeyHex);
+    // Parse the binary DER string into an ASN.1 object structure.
+    const ans1 = forge.asn1.fromDer(publicKeyDer);
```

Comment:
I suggest parsing the public key with the strict ASN.1 parser, as we did for the SOD to be "trailing 0 bullet proof". For example:


```suggestion
    const asn1 = forge.asn1.fromDer(publicKeyDer, { parseAllBytes: false });
```